fix for non-sparc sun machines

bahamas10 opened the pull request:
I've been chasing a bug that started in rubywarden: https://github.com/jcs/rubywarden/issues/71

It led me to rubeepass: https://gitlab.com/mjwhitta/rubeepass/issues/3

Which led me here.

I'm testing on a [SmartOS](https://smartos.org) machine, which uses the [illumos](http://illumos.org) kernel, which is a fork from Open Solaris (so it looks like a Sun machine).

```
$ uname -a
SunOS arbiter.rapture.com 5.11 joyent_20180830T001556Z i86pc i386 i86pc Solaris
```

When I run the tests on the latest master branch I see:

```
$ rake test
/opt/local/bin/ruby22 -I"lib:test"  "/opt/local/lib/ruby/2.2.0/rake/rake_test_loader.rb" "test/salsa20_test.rb" 
Run options: --seed 6784

# Running:

.....F........

Finished in 0.311694s, 44.9159 runs/s, 118.7062 assertions/s.

  1) Failure:
Salsa20Test#test_salsa20_keystream [/home/dave/dev/salsa20-ruby/test/salsa20_test.rb:10]:
--- expected
+++ actual
@@ -1 +1 @@
-"@\x8D\x94\xF48f9Z)\e\xBD\xB8?\xCC\xEC\xD6g\xB3;\xC7ev\v\xCA]\xEE\x19I;\xA2<\\^\xCEFQn\x94B{+\x06\xE2\x85\x9F\xEC\xBBp@\xA4\x8F\xD8~\xD3\x12\x197\f\xD7'\x8C\xC8\xEF\xFC"
+"\x1Fp\x05;\xE5\xEE{%\x06\x12\xD8\xB1\xEB\x9C\xD7,\xEF=|\x89\x9C\x96%\xCE-\xF4\xE3hgMUq\xD0\xFCB\x97N\xC9_4\xE3\xFA\xA4\xC9\x17S\x8Es\xEBe\xC5\xF4;\xD7F\xC2\x04X\x18\xA7h\x06\xD46"


14 runs, 37 assertions, 1 failures, 0 errors, 0 skips
rake aborted!
Command failed with status (1): [ruby -I"lib:test"  "/opt/local/lib/ruby/2.2.0/rake/rake_test_loader.rb" "test/salsa20_test.rb" ]
/opt/local/bin/rake22:33:in `<main>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

I traced this to an error in how my architecture was being determined.  This patch will correctly identify and separate sparc machines from sun machines.

With this patch:

```
$ rake test
/opt/local/bin/ruby22 -I"lib:test"  "/opt/local/lib/ruby/2.2.0/rake/rake_test_loader.rb" "test/salsa20_test.rb" 
Run options: --seed 39109

# Running:

..............

Finished in 0.001385s, 10108.7850 runs/s, 26716.0746 assertions/s.

14 runs, 37 assertions, 0 failures, 0 errors, 0 skips
```

I was also able to test what type of machine it was determining I had with this test program:

```
$ cd ext/salsa20_ext
$ cat test.c 
#include "ecrypt-sync.h"
#include <stdio.h>

int main() {
#if defined(ECRYPT_LITTLE_ENDIAN)
printf("ECRYPT_LITTLE_ENDIAN\n");
#elif defined(ECRYPT_LITTLE_ENDIAN)
printf("ECRYPT_LITTLE_ENDIAN\n");
#else
printf("UNKNOWN\n");
#endif
}
$ cc test.c salsa20.c 
$ ./a.out 
ECRYPT_LITTLE_ENDIAN
```

bahamas10 committed:
better determine machine endianess

- fixes non-sparc sun machines

dubek commented:
Thanks @bahamas10 for looking into this and contribution a fix. The patch looks good, but before I merge it I have a few questions:

1. In your test program both clauses check and print `ECRYPT_LITTLE_ENDIAN`. I think that's a bug.
2. The `ecrypt-config.h` is taken as-is from djb's salsa20 code ( https://cr.yp.to/snuffle/ecrypt-config.h ). I find it weird that his code fails on some machines (though it is possible). If you have an idea, let me know.
3. According to the `uname` output your machine is an `i386` machine. Why doesn't the compile define the `i386` setting? Which compiler is this?

bahamas10 commented:
### 1.

Oh wow, I messed that up ha.  I modified it and it still gives the output expected :p good catch!

```
[root@rubeepass ~/salsa20-ruby/ext/salsa20_ext]# cat test.c  
#include "ecrypt-sync.h"
#include <stdio.h>

int main() {
#if defined(ECRYPT_LITTLE_ENDIAN)
        printf("ECRYPT_LITTLE_ENDIAN\n");
#elif defined(ECRYPT_BIG_ENDIAN)
        printf("ECRYPT_BIG_ENDIAN\n");
#else
        printf("UNKNOWN\n");
#endif
}
[root@rubeepass ~/salsa20-ruby/ext/salsa20_ext]# cc test.c salsa20.c 
[root@rubeepass ~/salsa20-ruby/ext/salsa20_ext]# ./a.out 
ECRYPT_LITTLE_ENDIAN
```

Without the patch installed

```
[root@rubeepass ~/salsa20-ruby/ext/salsa20_ext]# cc test.c salsa20.c 
[root@rubeepass ~/salsa20-ruby/ext/salsa20_ext]# ./a.out 
ECRYPT_BIG_ENDIAN
```

### 2.

From https://cr.yp.to/salsa20.html

> History: I introduced Salsa20 in March 2005. It is a refinement of Salsa10, which I introduced in November 2004. 

A lot has happened in the world of Sun and the OpenSolaris operating system... and the fork as illumos.  I'm not surprised that this logic doesn't hold up any more.  However, I think conflating sun with sparc is the problem here.

### 3.

```
[root@rubeepass ~]# cc --version
cc (GCC) 4.9.4
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

[root@rubeepass ~]# arch
i86pc
[root@rubeepass ~]# ruby -e 'p RUBY_PLATFORM'
"x86_64-solaris2.11"
```

I don't know to answer your question exactly... I can only show you what I'm running with here.

---

If you would like to mess around on this machine to get a feel for it, I've added your SSH keys from GitHub to a test zone I've created... you should be able to login with:

    ssh 72.2.114.201 -l root

Feel free to login and mess around and don't worry about breaking anything.

dubek commented:
Thanks.

I won't have time to mess with that machine this weekend, maybe sometime next week; same for pushing a new gem version. 

If you want to look in the meantime, in https://github.com/cr-marcstevens/sha1collisiondetection/blob/master/lib/sha1.c they do a bunch of checks like `#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__` which are nice (clear) - not sure if they solve the problem on your machine. But I'm also OK with your current patch.

bahamas10 commented:
Sounds good! I took a look at the logic you sent in that link.  I added to the top of `ecrypt-config.h` as a sort of "check this first" thing, with a fallback method on the current best-guess method.

``` diff
[root@rubeepass ~/salsa20-ruby-dave]# git diff
diff --git a/ext/salsa20_ext/ecrypt-config.h b/ext/salsa20_ext/ecrypt-config.h
index a37e5e2..4b232f7 100644
--- a/ext/salsa20_ext/ecrypt-config.h
+++ b/ext/salsa20_ext/ecrypt-config.h
@@ -9,9 +9,31 @@
 
 /* Guess the endianness of the target architecture. */
 
+/* First try to look for precompiler hints */
+#if defined(__BYTE_ORDER__) && defined(__ORDER_LITTLE_ENDIAN__) \
+    && defined(__ORDER_BIG_ENDIAN__)
+
+#pragma message "__BYTE_ORDER__ defined"
+
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#pragma message "__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN"
+#define ECRYPT_LITTLE_ENDIAN
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+#pragma message "__BYTE_ORDER__ == __ORDER_BIG_ENDIAN"
+#define ECRYPT_BIG_ENDIAN
+#endif
+
+#endif
+
+/* No precompiler hints, fallback to best guess method */
+
 /* 
  * The LITTLE endian machines:
  */
+#if !defined(ECRYPT_BIG_ENDIAN) && !defined(ECRYPT_LITTLE_ENDIAN)
+
+#pragma message "Legacy Endian Lookup"
+
 #if defined(__ultrix)           /* Older MIPS */
 #define ECRYPT_LITTLE_ENDIAN
 #elif defined(__alpha)          /* Alpha */
@@ -52,6 +74,8 @@
 #define ECRYPT_UNKNOWN
 #endif
 
+#endif /* if !defined(ECRYPT_LITTLE_ENDIAN) ... */
+
 /* ------------------------------------------------------------------------- */
 
 /*
```

I added a lot of `pragma messages` to ensure that it was working as expected.  Sure enough, the logic all works fine on my machine!

```
../../../../ext/salsa20_ext/ecrypt-config.h:16:9: note: #pragma message: __BYTE_ORDER__ defined
 #pragma message "__BYTE_ORDER__ defined"
         ^
../../../../ext/salsa20_ext/ecrypt-config.h:19:9: note: #pragma message: __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN
 #pragma message "__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN"
```

If that diff looks good I can push it to this PR (without the debug messages).  Basically it uses `__BYTE_ORDER__` and friends if they are available, and fallsback to the current method if not.

dubek commented:
The logic in your diff looks great. Please push this into the PR and I'll merge and try to push a new gem version soon (haven't done it in months, I'll need to re-learn...).

bahamas10 commented:
great! I force-pushed a new single commit that incorporates 1. precompiler hints for machine endianess and 2. non sparc sun machine fixes for older compilers.

dubek commented:
Thank you @bahamas10 for your contribution.

This is now available in salsa20 gem version 0.1.3 (published to Rubygems).

bahamas10 commented:
awesome thank you!
